src/testdrive/src/action/kafka/ingest.rs

```diff
@@ -108,16 +117,56 @@ impl Action for IngestAction {
     fn redo(&self, state: &mut State) -> Result<(), String> {
         enum Encoder {
             Avro {
-                key_encode: Option<(Schema, i32)>,
-                value_schema: Schema,
-                value_schema_id: i32,
+                schema: Schema,
+                schema_id: i32,
             },
             Proto {
                 parser: &'static dyn Fn(&str) -> Result<protobuf::DynMessage, failure::Error>,
                 validator: &'static dyn Fn(&[u8]) -> Result<Box<dyn fmt::Debug>, failure::Error>,
             },
             Bytes,
+            Text,
         }
+        fn get_encoder(
```

Comment:
This would be a nice place to use a closure, as you can close over a bunch of variables rather than taking them as parameters:

```rust
        let mut make_encoder = |format, typ| match format {
            &Format::Avro { ref schema } => {
                let schema_id = if self.publish {
                    let ccsr_subject = format!("{}-{}", topic_name, typ);
                    let schema_id = state.tokio_runtime.block_on(
                        state
                            .ccsr_client
                            .publish_schema(&ccsr_subject, &schema)
                            .map_err(|e| format!("schema registry error: {}", e)),
                    )?;
                    schema_id
                } else {
                    1
                };
                let schema = interchange::avro::parse_schema(&schema)
                    .map_err(|e| format!("parsing avro schema: {}", e))?;
                Ok(Encoder::Avro { schema, schema_id })
            }
            &Format::Proto { ref message } => match message.as_str() {
                ".Struct" => Ok(Encoder::Proto {
                    parser: &protobuf::json_to_protobuf::<Struct>,
                    validator: &protobuf::decode::<Struct>,
                }),
                ".Batch" => Ok(Encoder::Proto {
                    parser: &protobuf::json_to_protobuf::<Batch>,
                    validator: &protobuf::decode::<Batch>,
                }),
                _ => Err(format!("unknown testdrive protobuf message: {}", message)),
            },
            &Format::Bytes => Ok(Encoder::Bytes),
            &Format::Text => Ok(Encoder::Text),
        };
        let value_encoder = make_encoder(&self.format, "value")?;
        let key_encoder = match &self.key_format {
            None => None,
            Some(f) => Some(make_encoder(f, "key")?),
        };
```